### PR TITLE
Feedback and star-rating module for sessions/speakers.

### PR DESCRIPTION
**Pull request by radi-cho**
More info: #615

### CODE REVIEW

**Comment by radi-cho** [comment]
@abraham Any progress with the merging?

**Comment by abraham** [comment]
@radi-cho thanks for the reminder. I'll take a look this weekend.

File: firestore.rules [hunk 1]
```diff
@@ -153,6 +157,10 @@ service cloud.firestore {
       allow write: if false;
     }
 
+    match /generatedSpeakers/{session}/feedback/{userId} {
```
Comment (abraham) [inline firestore.rules]: We can't assume that `generatedSpeakers` collection and subcollections won't randomly get deleted and regenerated so nothing should get written to them.

File: src/elements/dialogs/speaker-details.html [hunk 2]
```diff
@@ -128,6 +133,11 @@ <h3>{$ speakerDetails.sessions $}</h3>
           </template>
         </div>
 
+        <div class="additional-sections">
+          <h3>{$ feedback.headline $}</h3>
+          <feedback-block collection="generatedSpeakers" db-item="[[speaker.id]]"></feedback-block>
```
Comment (abraham) [inline src/elements/dialogs/speaker-details.html]: `generatedSpeakers` usage again.

File: src/elements/feedback.html [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+          ga-event-action="delete feedback"
+          ga-event-label$="delete the feedback record"
+        >
+          {$ feedback.deleteFeedback $}
+        </paper-button>
+      </div>
+    </template>
+
+    <script>
+      class Feedback extends ReduxMixin(Polymer.Element) {
```
Comment (abraham) [inline src/elements/feedback.html]: There is a lot of logic and methods in this class, I wonder if there might be some simplification that could happen.

File: src/elements/feedback.html [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+        static get is() {
+          return 'feedback-block';
+        }
+
+        connectedCallback() {
+          super.connectedCallback();
+          // Listen for previousFeedback state changes
+          // eslint-disable-next-line no-undef
+          store.subscribe(() => this._updateFeedbackState());
```
Comment (abraham) [inline src/elements/feedback.html]: I this `store` usage (and the `_updateFeedbackState` use) will make #579 more difficult. Is there an easier way to do? Maybe with `statePath` somehow?

**Comment by abraham** [comment]
Should there be a 48 hour limit on the review window? There should probably be a time after which reviews can't be left.

File: src/elements/feedback.html (hunk 3 above)
Comment (radi-cho) [reply src/elements/feedback.html]: If we think about it, the module actually handles a lot of stuff..
- Makes sure the user is authenticated and checks if it is appropriate to display the feedback box;
- Fetches previous feedback;
- Maintains the fetched data in the redux state;
- Updates properly all info when the user switches between sessions;
- Submits the feedback to the db;

File: src/elements/feedback.html (hunk 4 above)
Comment (radi-cho) [reply src/elements/feedback.html]: I think I should re-do this PR some day, for Polymer 3.

**Comment by radi-cho** [comment]
### 1. Feedback time window.
I think the GDGs which are using hoverboard should decide on that separately.
Maybe reviews after some time just doesn't matter. I personally think that day or two after the DevFest event the most of the visitors willing to give feedback are gone, so it doesn't matter.

### 2. Speaker reviews
Basically, the same as `[1]`. Different GDGs may use the provided feedback differently. The main problem is that one speaker may talk in 2+ sessions and one session may be presented by 2+ speakers.

### 3. Polymer 3 [IMPORTANT]; Redux store subscribe; star-rating component.
I think it will be easier if I just rewrite the hole thing for Polymer 3 extending the `polymer-3` branch along with logic simplification and dependency compatibility.

> It will be really hard to maintain all the functionalities of the feedback module, make them compatible with sites using Polymer 2, but also to make it upgradeble to Polymer 3 in one PR.

### 4. The rest.
I've fixed the rest. Please check out the replies to some of the code comments.

**Comment by abraham** [comment]
1. I'm suggesting closing the review window mostly because reviews should happen pretty quick and after that leaving reviews open is a spam/abuse risk. I think this should default to closing 7 days after an event.

2. I think the reviews should be per session. If a speaker has multiple sessions they should have reviews for each as the sessions might be very different. If one is meh and the other is really good the speaker probably wants to know that. If multiple people present in a single session the session reviews should apply to both of them equally.

3. Are you saying you'd like to refactor this and merge it after the polymer 3 upgrade? I would like to complete that in the next month, but it's also been pending for a while.

I just thought of an additional question. Are there sessions that visitors should not be able to review? Like check-in/lunch/after party? Should there be a way to disable reviews on specific sessions?

File: data/resources.json [hunk 5]
```diff
@@ -23,6 +23,8 @@
   "country": "Country",
   "time": "Time",
   "faq": "/data/faq.md",
+  "cod": "/data/cod.md",
```
Comment (abraham) [inline data/resources.json]: `cod` should be removed. It was replaced with `coc` below.

File: firestore.rules [hunk 6]
```diff
@@ -153,6 +157,10 @@ service cloud.firestore {
       allow write: if false;
     }
 
+    match /speakers/{session}/feedback/{userId} {
```
Comment (abraham) [inline firestore.rules]: This `match` is included twice.

**Comment by radi-cho** [comment]
1. I am going to add a time window.
2. I think it is fine to remove the speaker review, but still there may be some special occasions when it is helpful to exist. Different GDGs can add it separately, it is 2 lines of code anyway.
3. Yes. I would like to refactor the feedback block ..`and merge it after the polymer 3 upgrade`. The changes should be ready by the end of September.
4. I think that some sessions may not matter, but someone may leave feedback for the food for the lunch session or the drinks at the party..

**Comment by radi-cho** [comment]
What is happening with this?

**Comment by abraham** [comment]
I definitely want to get this merged but I've been a bit short on time. I'll try to get to it soon.

**Pull request closed without merging**